File: cpp/hap_cpp2/src/session_crypto.cpp
#include "session_crypto.h"

#define MAC_SIZE 4

void SessionCrypto::encrypt(const uint8_t* in, int in_len, uint8_t*& out, int& out_len)
{
    free(write_buf);
    int offset = 0;
    int to_len = 0;
    out = 0;
    while (offset < in_len)
    {
        int chunk_len = in_len - offset;
        if (chunk_len > MAX_ENCRYPTED_LENGTH)
        {
            chunk_len = MAX_ENCRYPTED_LENGTH;
        }
        printf ("session encrypt: processing chunk len = %d\n", chunk_len);
        encryptChunk(&in[offset], chunk_len, out, to_len);
        offset += chunk_len;
    }

    out_len = to_len;
    write_buf = out;
}

void SessionCrypto::encryptChunk(const uint8_t* from, int from_len, uint8_t*& to, int& to_len)
{
    uint8_t additionalData[2];
    additionalData[0] = from_len & 0xff;
    additionalData[1] = from_len >> 8;

//    uint8_t* nonce = Pack.longToLittleEndian(outboundBinaryMessageCount++);
//    uint8_t* ciphertext = new ChachaEncoder(write_key, nonce).encodeCiphertext(chunk, additionalData);

    const uint8_t* ciphertext = from;
    int ciphertext_len = from_len + MAC_SIZE;  //TODO has to have real ciphertext len (from_len + 16)

    to = (uint8_t*) realloc (to, to_len + ciphertext_len + 2);
    to[to_len] = additionalData[0];
    to[to_len + 1] = additionalData[1];
    memcpy (&to[to_len + 2], ciphertext, from_len);  //TODO has to be ciphertext_len
    //fake auth array of 16 elements
    for (int i = 0; i < MAC_SIZE; ++i)
    {
        to[to_len + from_len + 2 + i] = i+1;
    }
    to_len += ciphertext_len + 2;
}


void SessionCrypto::decrypt(const uint8_t* from, int from_len, uint8_t*& to, int& to_len)
{
    free (read_buf);
    to_len = 0;
    to = 0;
    int offset = 0;
    while (offset < from_len)
    {
        int chunk_len = ((from[offset + 1] & 0xff) << 8) + (from[offset + 0] & 0xff);
        printf ("session decrypt: processing chunk len = %d\n", chunk_len);
        decryptChunk(&from[offset + 2], chunk_len, to, to_len);
        offset += chunk_len + 2 + MAC_SIZE;
    }
    read_buf = to;

}

void SessionCrypto::decryptChunk (const uint8_t* from, int from_len, uint8_t*& to, int& to_len)
{
    static uint8_t mac[MAC_SIZE];  //TODO depends on input parameters of our real decrypt code.
    memcpy (mac, &from[from_len], MAC_SIZE); //so may not need this

//    uint8_t additionalData[2];
//    additionalData[0] = from_len & 0xff;
//    additionalData[1] = from_len >> 8;

    //    byte[] nonce = Pack.longToLittleEndian(inboundBinaryMessageCount++);
//    plaintextChunk = new ChachaDecoder(read_key, nonce).decodeCiphertext(mac, additionalData, cipherchunk);

    const uint8_t* plaintextChunk = from;
    to = (uint8_t*) realloc (to, to_len + from_len);
    memcpy (&to[to_len], plaintextChunk, from_len); //TODO plainchunk len
    to_len += from_len; //TODO plainchunk len
}

void SessionCrypto::setKeys(uint8_t* write_key, uint8_t* read_key)
{
    this->write_key = write_key;
    this->read_key = read_key;
}

File: hap_mininal11/proj/main/SessionCrypto.java
package main;

import java.io.IOException;

import bouncy.Pack;
import crypto.ChachaDecoder;
import crypto.ChachaEncoder;

public class SessionCrypto
{
    private static final int MAX_ENCRYPTED_LENGTH = 1024;
    int inboundBinaryMessageCount = 0;
    int outboundBinaryMessageCount = 0;
    private AccessoryCfg cfg;
    
    public SessionCrypto(AccessoryCfg cfg)
    {
        this.cfg = cfg;
    }
    
    private byte[] addChunk(byte[] from, int len, byte[] to, int offset) throws Exception
    {
        byte[] nonce = Pack.longToLittleEndian(outboundBinaryMessageCount++);

        byte[] additionalData = new byte[2];

        byte[] chunk;
        if (offset == 0 && len < MAX_ENCRYPTED_LENGTH)
        {
            chunk = from;
        }
        else
        {
            chunk = new byte[len];
            for (int i = 0; i < len; ++i)
            {
                chunk[i] = from[i + offset];
            }
        }

        additionalData[0] = (byte) (chunk.length & 0xff);
        additionalData[1] = (byte) (chunk.length >> 8);

        byte[] ciphertext = new ChachaEncoder(cfg.getWriteKey(), nonce).encodeCiphertext(chunk, additionalData);

        // ///////////////
        byte[] merge = new byte[to.length + ciphertext.length + 2];

        int i;
        for (i = 0; i < to.length; ++i)
        {
            merge[i] = to[i];
        }

        merge[i] = additionalData[0];
        merge[i + 1] = additionalData[1];

        for (i = 0; i < ciphertext.length; ++i)
        {
            merge[i + 2 + to.length] = ciphertext[i];
        }
        
        return merge;
    }
    
    public byte[] decrypt(byte[] msg)
    {
        byte[] mac = new byte[16];
        byte[] ciphertext = new byte[msg.length - 16];
        System.arraycopy(msg, 0, ciphertext, 0, msg.length - 16);
        System.arraycopy(msg, msg.length - 16, mac, 0, 16);
        int msglen = msg.length - 16;
        byte[] additionalData = new byte[2];
        additionalData[0] = (byte) (msglen & 0xff);
        additionalData[1] = (byte) (msglen >> 8);

        try
        {
            byte[] nonce = Pack.longToLittleEndian(inboundBinaryMessageCount++);
            return new ChachaDecoder(cfg.getReadKey(), nonce).decodeCiphertext(mac, additionalData, ciphertext);
        }
        catch (IOException e)
        {
            throw new RuntimeException(e);
        }
    }

    public byte[] encrypt(byte[] plaintext) throws Exception
    {
        int offset = 0;
        byte[] result = new byte[0];
        while (offset < plaintext.length)
        {
            int len = plaintext.length - offset;
            if (len > MAX_ENCRYPTED_LENGTH)
            {
                len = MAX_ENCRYPTED_LENGTH;
            }
            result = addChunk(plaintext, len, result, offset);
            offset += len;
        }

        return result;
    }
   
 
}
